File: src/algorithms.rs
pub fn distance(matrix: Vec<Vec<f32>>, city1: City, city2: City) -> f32{
    if ((matrix[city1.city_number])[city2.city_number] = -1.0)
    maximum_distance * natural_distance(matrix, city1, city2)
    else
    (matrix[city1.city_number])[city2.city_number]



}

fn maximum_distance(matrix: Vec<Vec<f32>>) -> f32{
    let counter = -1.0;
    for x in 0..matrix.len()
    for y in 0..matrix.len()
    if (counter < (matrix[x])[y])
    counter = (matrix[x])[y])
    counter
}

fn natural_distance(matrix: Vec<Vec<f32>>, city1: City, city2: City) -> f32{
    let r = 6373000.0;

    r * 2 * atan(sqrt(a(city1, city2)),sqrt(1 − a(city1, city2)));
}

fn a(city1: City, city2: City){
    sin(powi((city1.x - city2.x)/2 , 2)) + cos(city1.x) * cos(city2.x) *
    sin(powi((city1.y - city2.y)/2 , 2)) * 1.2
}
// fn convert_radians(g: f32) -> f32{
//     g * f32::consts::PI / 180
// }

// Path: 1,2,3,4,5,6,7,163,164,165,168,172,327,329,331,332,333,489,490,491,492,493,496,653,654,656,657,661,815,816,817,820,823,871,978,979,980,981,982,984
// Evaluation: 3305585.454990047
//
// • Y otra de 150:
//
// Path: 1,2,3,4,5,6,7,8,9,11,12,14,16,17,19,20,22,23,25,26,27,74,75,151,163,164,165,166,167,168,169,171,172,173,174,176,179,181,182,183,184,185,186,187,297,326,327,328,329,330,331,332,333,334,336,339,340,343,344,345,346,347,349,350,351,352,353,444,483,489,490,491,492,493,494,495,496,499,500,501,502,504,505,507,508,509,510,511,512,520,652,653,654,655,656,657,658,660,661,662,663,665,666,667,668,670,671,673,674,675,676,678,815,816,817,818,819,820,821,822,823,825,826,828,829,832,837,839,840,871,978,979,980,981,982,984,985,986,988,990,991,995,999,1001,1003,1004,1037,1038,1073,1075
// Evaluation: 6152051.625245280

File: src/proyecto1/Algorithms.java
package proyecto1;

import java.lang.Math;
import java.util.ArrayList;

public class Algorithms {
	
	public static double calculateDistance(Connections connections, City city1, City city2, double max) {
		double distance = connections.getDistance(city1.getCityNumber(), city2.getCityNumber());
		if(distance == -1.0)
			distance = max * naturalDistance(connections, city1, city2);
		return distance;
	}

	private static double naturalDistance(Connections connections, City city1, City city2) {
		double r = 6373000.0;
		
		return r * 2 * Math.atan2(Math.sqrt(a(city1, city2)),Math.sqrt(1-a(city1,city2)));
	}

	//private static double a(City city1, City city2) {
	//	double a = Math.sin( Math.pow(((city1.getX() - city2.getX())/2),2) ) + Math.cos(city1.getX()) * Math.cos(city2.getX())
	//			* Math.sin( Math.pow(((city1.getY() - city2.getY())/2),2) );
	//	return a;
	//}
	
	private static double a(City city1, City city2) {
		double a = Math.sin( Math.pow(((rad(city1.getX()) - rad(city2.getX()))/2),2) )
				+ Math.cos(rad(city1.getX())) * Math.cos(rad(city2.getX()))
				* Math.sin( Math.pow(((rad(city1.getY()) - rad(city2.getY()))/2),2) );
		return a;
	}
	
	private static double rad(double g) {
		return g * Math.PI / 180;
	}
	

	public static boolean esFactible(int[] elements,Connections connections) {
		for(int i = 0; i < elements.length - 1;i++) {
			if(connections.getDistance(elements[i], elements[i+1]) == -1.0)
				return false;
		}
		return true;
		
	}
	
	public static double normalizer(int[] elements, Connections connections) {
		ArrayList<Double> normalize = new ArrayList<Double>();
		double currentMax;
		for(int i = 0; i < elements.length; i++) {
			currentMax = 0;
			
			for(int j = 0; j < elements.length; j++) {
				if(j <= i || i == j+1 )
					continue;
				if(currentMax < connections.getDistance(elements[i], elements[j])) 
					currentMax = connections.getDistance(elements[i], elements[j]);		
			}	
			normalize.add(currentMax);
		}
		
		double counter = 0.0;
		for(Double d: normalize)
			counter += d;
		
		return counter;
	}
	
	public static double funcionDeCosto(int[] permutation, Connections connections, World world, double normal) {
		double total = 0.0;
		for(int i = 0; i < permutation.length; i++) 
			for(int j = 0; j < permutation.length; j++) {
				total += calculateDistance(connections, world.getCity(permutation[i]),
											world.getCity(permutation[j]), connections.findMax(permutation));
			}

		return total/normal;
	}
	
}
